{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyNEy184jbU4UivXE/vYVgWp"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# BUSCO"],"metadata":{"id":"9NK86DoPw1B1"}},{"cell_type":"markdown","source":["## What is BUSCO ?\n","BUSCO provides a quantitative assessment of the completeness in terms of expected gene content of a genome assembly using a database containing orthologs genes."],"metadata":{"id":"UDRdxVKBxJ6E"}},{"cell_type":"markdown","source":["# Running BUSCO\n"],"metadata":{"id":"rBjFuAKTel7c"}},{"cell_type":"code","source":["$ busco --list-datasets ## List all the available dataset, the one interesting to us are fungi_odb10, and to be more precise : ascomycota_odb10,leotiomycetes_odb10"],"metadata":{"id":"bgMgxLWDeot4"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Create a conda environnment for busco"],"metadata":{"id":"CDW95jG8ycn3"}},{"cell_type":"code","source":["$ conda create -n busco -c conda-forge -c bioconda busco=5.6.0\n","$ source ~/.bashrc\n","$ conda activate busco"],"metadata":{"id":"hCzWOCrwfl7O"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#!/bin/bash \n","\n","# Activate the conda environment\n","conda activate busco\n","\n","# Create directory for BUSCO results\n","mkdir -p \"/bigvol/omion/08-QualityAssemblies/BUSCO\"\n","\n","# Loop through all assembly files\n","for i in /bigvol/omion/All_assemblies/Gd*/*.fasta; do\n","    if [ -f \"$i\" ]; then\n","        gd_part=$(basename $(dirname \"$i\"))\n","        assembly_name=$(basename \"$i\" .fasta)\n","\n","        # Create the output directory path\n","        output_dir=\"/bigvol/omion/08-QualityAssemblies/BUSCO/${gd_part}\"\n","\n","        # Run BUSCO\n","        busco -i \"$i\" -m genome -l leotiomycetes_odb10 -o \"${output_dir}\" -f\n","    fi\n","done\n","\n","# Deactivate the conda environment\n","conda deactivate\n"],"metadata":{"id":"fKsLQ31tQALA"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Creation of BUSCO plots"],"metadata":{"id":"aorxARipucy3"}},{"cell_type":"markdown","source":["## Code to moove the short summary from busco to their respective busco_summaries folder for the creation of BUSCO plots"],"metadata":{"id":"D3yEg2UcvDex"}},{"cell_type":"code","source":["#!/bin/bash\n","\n","# Create a folder to put all the busco short summaries\n","mkdir -p busco_summaries\n","\n","# Traverse all directories except for the \"logs\" directory\n","for dir in /bigvol/omion/08-QualityAssemblies/BUSCO/Gd*/busco_Gd*/; do\n","    if [ \"$dir\" != \"/bigvol/omion/08-QualityAssemblies/BUSCO/Gd*/busco_Gd*/logs/\" ]; then\n","        dir_name=$(basename \"$dir\")\n","        echo \"Processing directory $dir_name...\"\n","\n","        # Create a directory for each main directory\n","        mkdir -p \"busco_summaries/$dir_name\"\n","\n","        # Copy all short_summary*.txt files to busco_summaries/main_directory_name/\n","        cp \"$dir\"**/short_summary*.txt \"./busco_summaries/$dir_name/\"\n","    fi\n","done\n"],"metadata":{"id":"kfNDkWKrt8x2"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Code to create BUSCO plots"],"metadata":{"id":"0MQSN_ixu_gH"}},{"cell_type":"code","source":["#!/bin/bash\n","\n","for dir in ./busco_summaries/busco_Gd*; do\n","  # Execute the generate_plot.py script with the current directory\n","  generate_plot.py -wd \"$dir\"\n","done\n"],"metadata":{"id":"iCMoM-E3uAKF"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Obtain lenght of contigs\n","\n","Quast does not output by default the lenght of the contigs, they will only be available in the html viewver but one by one, and not in group, so to obtain directly the lengh of all the contigs of a sample we can execute :  "],"metadata":{"id":"HocTsxEtYyoV"}},{"cell_type":"code","source":["#!/bin/bash\n","\n","# Loop over each file matching the pattern filtered_hypo*\n","for file in /bigvol/omion/05-Polishing/HyPo/Gd*/hypo_Gd*.fasta; do\n","  # Extract the specific part of the filename (Gd1111)\n","  prefix=$(echo \"$file\" | sed -E 's/filtered_hypo_(Gd[^_]+)_.*/\\1/')\n","\n","  # Process the file with awk\n","  awk -v prefix=\"$prefix\" '\n","  /^>/ {\n","    if (seq) {\n","      print prefix \"\\t\" substr(name, 2) \"\\t\" length(seq);\n","    }\n","    name = $0;\n","    seq = \"\";\n","  }\n","  !/^>/ {\n","    seq = seq $0;\n","  }\n","  END {\n","    if (seq) {\n","      print prefix \"\\t\" substr(name, 2) \"\\t\" length(seq);\n","    }\n","  }' \"$file\" >> contig_lengths.tsv\n","done\n"],"metadata":{"id":"f9vN-0FTZHpU"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# QUAST\n"],"metadata":{"id":"ZnSAyeSbILNH"}},{"cell_type":"markdown","source":["## What is QUAST ?\n","QUAST provides an analysis about the quality of assemblies, notably at the level of the contiguity  (N50,N90, nÂ° of contigs, largest contig, total lenght in bp...)."],"metadata":{"id":"azjFfAYRw36O"}},{"cell_type":"markdown","source":["# Running QUAST\n"],"metadata":{"id":"ILzLjmgRx-zH"}},{"cell_type":"code","source":["#!/bin/bash  Place yourself in the folder /bigvol/omion/All_assemblies\n","\n","# Define the isolates\n","isolates=(\"Gd1111\" \"Gd2185\" \"Gd2407\" \"Gd267\" \"Gd442\" \"Gd4985\" \"Gd4986\" \"Gd614\" \"Gd708\" \"Gd994\")\n","\n","# Loop through each isolate\n","for isolate in \"${isolates[@]}\"; do\n","    # Define the directories and files for the current isolate\n","    assembly_guppy=\"/bigvol/omion/All_assemblies/${isolate}/assembly_${isolate}_Guppy_modbasecalling.fasta\"\n","    assembly_dorado=\"/bigvol/omion/All_assemblies/${isolate}/assembly_${isolate}_dorado_modbasecalling.fasta\"\n","    hypo_guppy=\"/bigvol/omion/All_assemblies/${isolate}/hypo_${isolate}_Guppy_modbasecalling.fasta\"\n","    hypo_dorado=\"/bigvol/omion/All_assemblies/${isolate}/hypo_${isolate}_Dorado_modbasecalling.fasta\"\n","    final=\"/bigvol/omion/All_assemblies/${isolate}/final${isolate#\"Gd\"}.fasta\"\n","    filtered_hypo_dorado=\"/bigvol/omion/All_assemblies/${isolate}/filtered_hypo_${isolate}_Dorado_modbasecalling.fasta\"\n","\n","    # Create QUAST output directory for the current isolate\n","    output_dir=\"./Quast/${isolate}/\"\n","    mkdir -p \"$output_dir\"\n","\n","    # Run QUAST and store output in the specified directory\n","    quast -o \"$output_dir\" -l \"assembly_${isolate}_Guppy_modbasecalling,assembly_${isolate}_dorado_modbasecalling,hypo_${isolate}_Guppy_modbasecalling,hypo_${isolate}_Dorado_modbasecalling,final${isolate#\"Gd\"},filtered_hypo_${isolate}_Dorado_modbasecalling\" \"$assembly_guppy\" \"$assembly_dorado\" \"$hypo_guppy\" \"$hypo_dorado\" \"$final\" \"$filtered_hypo_dorado\" --fungus --fragmented --k-mer-stats --threads 40\n","done\n","\n","isolates=(\"Gd293\" \"Gd45\")\n","\n","# Loop through each isolate\n","for isolate in \"${isolates[@]}\"; do\n","    # Define the directories and files for the current isolate\n","    assembly_gup=\"/bigvol/omion/All_assemblies/${isolate}/assembly_${isolate}_Guppy_basecalling.fasta\"\n","    assembly_guppy=\"/bigvol/omion/All_assemblies/${isolate}/assembly_${isolate}_Guppy_modbasecalling.fasta\"\n","    assembly_dor=\"/bigvol/omion/All_assemblies/${isolate}/assembly_${isolate}_dorado_basecalling.fasta\"\n","    assembly_dorado=\"/bigvol/omion/All_assemblies/${isolate}/assembly_${isolate}_dorado_modbasecalling.fasta\"\n","    hypo_gup=\"/bigvol/omion/All_assemblies/${isolate}/hypo_${isolate}_Guppy_basecalling.fasta\"\n","    hypo_guppy=\"/bigvol/omion/All_assemblies/${isolate}/hypo_${isolate}_Guppy_modbasecalling.fasta\"\n","    hypo_dor=\"/bigvol/omion/All_assemblies/${isolate}/hypo_${isolate}_Dorado_basecalling.fasta\"\n","    hypo_dorado=\"/bigvol/omion/All_assemblies/${isolate}/hypo_${isolate}_Dorado_modbasecalling.fasta\"\n","    final=\"/bigvol/omion/All_assemblies/${isolate}/final${isolate#\"Gd\"}.fasta\"\n","    filtered_hypo_dorado=\"/bigvol/omion/All_assemblies/${isolate}/filtered_hypo_${isolate}_Dorado_modbasecalling.fasta\"\n","\n","\n","    # Create QUAST output directory for the current isolate\n","    output_dir=\"./Quast/${isolate}/\"\n","    mkdir -p \"$output_dir\"\n","\n","    # Run QUAST and store output in the specified directory\n","    quast -o \"$output_dir\" -l \"assembly_${isolate}_Guppy_basecalling,assembly_${isolate}_Guppy_modbasecalling,assembly_${isolate}_dorado_basecalling,assembly_${isolate}_dorado_modbasecalling,hypo_${isolate}_Guppy_basecalling,hypo_${isolate}_Guppy_modbasecalling,hypo_${isolate}_Dorado_basecalling,hypo_${isolate}_Dorado_modbasecalling,final${isolate#\"Gd\"},filtered_hypo_${isolate}_Dorado_modbasecalling\" \"$assembly_gup\" \"$assembly_guppy\" \"$assembly_dor\" \"$assembly_dorado\" \"$hypo_gup\" \"$hypo_guppy\" \"$hypo_dor\" \"$hypo_dorado\" \"$final\" \"$filtered_hypo_dorado\" --fungus --fragmented --k-mer-stats --threads 40\n","done\n"],"metadata":{"id":"2qaShnZUSkVG"},"execution_count":null,"outputs":[]}]}
